x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1117,6 +1165,58 @@ Collection<Object> createComponents(
         return components;
     }
 
+    private void applyPendingSecurityMigrations(
+        ClusterState clusterState,
+        FeatureService featureService,
+        SecurityIndexManager.State newState
+    ) {
+        if (clusterState.nodes().isLocalNodeElectedMaster() == false) {
+            return;
+        }
+
+        Integer nextMigrationVersion = SecurityMigrations.MIGRATIONS_BY_VERSION.higherKey(newState.migrationsVersion);
+
+        // Check if next migration that has not been applied is eligible to run on the current cluster
+        boolean eligibleSecurityMigrationExists = nextMigrationVersion != null
+            && isEligibleSecurityMigration(
+                loadIndexMappingVersion(SECURITY_MAIN_ALIAS, clusterState),
+                featureService,
+                clusterState,
+                SecurityMigrations.MIGRATIONS_BY_VERSION.get(nextMigrationVersion)
+            );
+
+        if (eligibleSecurityMigrationExists == false) {
+            // Reset retry counter if all eligible migrations have been applied successfully
+            nodeLocalMigrationRetryCount.set(0);
+        } else if (nodeLocalMigrationRetryCount.get() > MAX_SECURITY_MIGRATION_RETRY_COUNT) {
+            logger.warn("Security migration failed [" + nodeLocalMigrationRetryCount.get() + "] times, restart node to retry again.");
+            return;
+        }
+
+        if (eligibleSecurityMigrationExists && clusterReadyForSecurityMigration(newState)) {
```

Comment:
Why isn't the `clusterReadyForSecurityMigration` included in the `eligibleSecurityMigrationExists` above? I think they should be. The above checks if the index has the pre-requisite mapping and this checks if the index exists (with the expected index.format setting) and is available.

Reply:
There is a (theoretical) corner case where the migrations are not up to date (`eligibleSecurityMigrationExists = true`) and `clusterReadyForSecurityMigration(newState)` flips back and forth when the state changes. In this case we do not want to reset the `nodeLocalMigrationRetryCount` since all of the eligible migrations were never successfully applied.